File: scripts/deploy-space.js
/**
 * Script de déploiement du Space Hugging Face (convertisseurs 3D + Office).
 *
 * Utilisation:
 *   HF_TOKEN=... npm run deploy:space -- --space-id <utilisateur>/<space>
 *
 * Variables d'environnement requises:
 *   - HF_TOKEN: Token Hugging Face avec permissions "write"
 *
 * Options en ligne de commande:
 *   --space-id <username/space-name>  : ID personnalisé (défaut: <username>/solidworks-viewer)
 *   --private                         : Rendre le Space privé (défaut: public)
 *   --skip-files                      : Ne pas uploader les fichiers (seulement création)
 *
 * Si le Space existe déjà, la création est ignorée et les fichiers sont mis
 * à jour via un commit atomique (format NDJSON officiel du Hub).
 */

import { existsSync, readFileSync } from 'fs';
import { dirname, join, resolve } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));

// Configuration
const SPACE_NAME = 'solidworks-viewer';
const SPACE_SDK = 'docker';
const SPACE_HARDWARE = 'cpu-basic'; // cpu-basic, cpu-upgrade, t4-medium, a10g-small, a10g-large
const HF_API = 'https://huggingface.co/api';

/** Fichiers synchronisés vers le Space (commit atomique unique). */
const SPACE_FILES = [
  'Dockerfile',
  'requirements.txt',
  'app.py',
  'freecad_convert.py',
  'freecad_cad_convert.py',
  'README.md',
];

/**
 * Parse les arguments CLI.
 */
function parseArgs(argv = process.argv.slice(2)) {
  const options = {
    spaceId: null,
    private: false,
    skipFiles: false,
  };

  for (let i = 0; i < argv.length; i++) {
    if (argv[i] === '--space-id' && argv[i + 1]) {
      options.spaceId = argv[++i];
    } else if (argv[i] === '--private') {
      options.private = true;
    } else if (argv[i] === '--skip-files') {
      options.skipFiles = true;
    } else if (argv[i] === '--help' || argv[i] === '-h') {
      console.log(`
Usage: node deploy-space.js [options]

Options:
  --space-id <id>     Space ID (format: username/space-name)
  --private           Créer un Space privé
  --skip-files        Ne pas uploader les fichiers
  --help, -h          Afficher cette aide

Variables d'environnement:
  HF_TOKEN            Token Hugging Face (requis)
`);
      process.exit(0);
    }
  }

  return options;
}

function authHeaders() {
  return { Authorization: `Bearer ${process.env.HF_TOKEN}` };
}

/**
 * Récupère le username depuis l'API Hugging Face.
 */
async function getUsername() {
  let response;
  try {
    response = await fetch(`${HF_API}/whoami-v2`, { headers: authHeaders() });
  } catch {
    throw new Error('Connexion à Hugging Face impossible.');
  }

  if (response.status === 401) {
    throw new Error('Token HF_TOKEN invalide.');
  }
  if (!response.ok) {
    throw new Error('Impossible de récupérer les informations utilisateur');
  }

  const data = await response.json();
  return data.name;
}

/**
 * Vérifie si le Space existe déjà (évite un appel de création inutile).
 */
async function spaceExists(spaceId) {
  let response;
  try {
    response = await fetch(`${HF_API}/spaces/${spaceId}`, { headers: authHeaders() });
  } catch {
    throw new Error('Connexion à Hugging Face impossible.');
  }

  if (response.status === 401) {
    throw new Error('Token HF_TOKEN invalide.');
  }
  return response.ok;
}

/**
 * Crée le Space via l'API Hugging Face.
 * `name` ne contient que le nom du dépôt, le namespace va dans `organization`
 * (comme `huggingface_hub.create_repo`).
 */
async function createSpace(spaceId, username, isPrivate) {
  console.log(`\n🚀 Création du Space: ${spaceId}`);
  console.log(`   SDK: ${SPACE_SDK}`);
  console.log(`   Hardware: ${SPACE_HARDWARE}`);
  console.log(`   Visibilité: ${isPrivate ? 'privé' : 'public'}`);

  const slash = spaceId.indexOf('/');
  const name = slash === -1 ? spaceId : spaceId.slice(slash + 1);
  const organization = slash === -1 ? null : spaceId.slice(0, slash) || null;

  try {
    const response = await fetch(`${HF_API}/repos/create`, {
      method: 'POST',
      headers: { ...authHeaders(), 'Content-Type': 'application/json' },
      body: JSON.stringify({
        type: 'space',
        name,
        organization,
        sdk: SPACE_SDK,
        hardware: SPACE_HARDWARE,
        private: isPrivate,
      }),
    });

    if (response.status === 409) {
      console.log(`⚠️  Le Space ${spaceId} existe déjà`);
      return false;
    }

    if (!response.ok) {
      const error = await response.text();
      throw new Error(`Erreur API: ${response.status} - ${error}`);
    }

    console.log(`✅ Space créé avec succès`);
    return true;
  } catch (error) {
    console.error('❌ Erreur lors de la création du Space:', error.message);
    throw error;
  }
}

/**
 * Pousse tous les fichiers du Space en un seul commit atomique
 * (format NDJSON officiel : en-tête + un objet `file` base64 par fichier).
 */
async function uploadSpaceFiles(spaceId, sourceDir) {
  console.log(`\n📁 Upload des fichiers depuis ${sourceDir}`);

  const operations = [];
  for (const file of SPACE_FILES) {
    const filePath = join(sourceDir, file);
    if (!existsSync(filePath)) {
      throw new Error(`Fichier source manquant: ${filePath}`);
    }
    console.log(`   📤 Commit: ${file}`);
    operations.push({
      key: 'file',
      value: { content: readFileSync(filePath).toString('base64'), path: file, encoding: 'base64' },
    });
  }

  const lines = [
    { key: 'header', value: { summary: 'Deploy converters (3D + Office)', description: '' } },
    ...operations,
  ];
  const body = lines.map((line) => JSON.stringify(line)).join('\n');

  let response;
  try {
    response = await fetch(`${HF_API}/spaces/${spaceId}/commit/main`, {
      method: 'POST',
      headers: { ...authHeaders(), 'Content-Type': 'application/x-ndjson' },
      body,
    });
  } catch {
    throw new Error('Connexion à Hugging Face impossible pendant l’upload.');
  }

  if (!response.ok) {
    const error = await response.text();
    throw new Error(`Erreur upload: ${response.status} - ${error}`);
  }

  console.log(`✅ Commit poussé (${operations.length} fichiers)`);
}

/**
 * Attend que le Space soit déployé.
 */
async function waitForDeployment(spaceId, timeout = 600000) {
  console.log(`\n⏳ Attente du déploiement (timeout: ${timeout / 60000}min)...`);

  const startTime = Date.now();
  const checkInterval = 10000; // 10 secondes

  while (Date.now() - startTime < timeout) {
    try {
      const response = await fetch(`${HF_API}/spaces/${spaceId}`, { headers: authHeaders() });

      if (!response.ok) {
        throw new Error(`Erreur status: ${response.status}`);
      }

      const data = await response.json();
      const runtime = data.runtime || {};
      const stage = runtime.stage || 'NO_APP_FILE';

      console.log(`   Status: ${stage}`);

      if (stage === 'RUNNING') {
        console.log(`✅ Space déployé et opérationnel!`);
        return true;
      }

      if (stage === 'RUNTIME_ERROR' || stage === 'BUILD_ERROR') {
        const detail = runtime.message || JSON.stringify(runtime);
        console.error(`❌ Erreur de déploiement: ${detail}`);
        return false;
      }
    } catch (error) {
      console.error(`   Erreur vérification status:`, error.message);
    }

    await new Promise((resolveTimer) => setTimeout(resolveTimer, checkInterval));
  }

  console.error(`⏰ Timeout atteint - le déploiement est toujours en cours`);
  return false;
}

/**
 * Fonction principale.
 */
async function main() {
  const options = parseArgs();

  // Afficher l'aide si demandée (avant vérification du token)
  if (process.argv.includes('--help') || process.argv.includes('-h')) {
    return; // Déjà affiché par parseArgs
  }

  console.log('🔧 Déploiement automatique du Space ENISE Converters (3D + Office)\n');

  // Vérification du token
  if (!process.env.HF_TOKEN) {
    console.error('❌ Variable HF_TOKEN manquante');
    console.error('   Usage: HF_TOKEN=your_token npm run deploy:space -- --space-id <user>/<space>');
    process.exit(1);
  }
  const sourceDir = join(__dirname, '..', 'space-huggingface');

  // Vérifier que les fichiers sources existent
  if (!existsSync(sourceDir)) {
    console.error(`❌ Dossier source introuvable: ${sourceDir}`);
    process.exit(1);
  }

  try {
    // Récupérer le username
    console.log('📋 Récupération des informations utilisateur...');
    const username = await getUsername();

    // Déterminer le spaceId
    const spaceId = options.spaceId || `${username}/${SPACE_NAME}`;
    console.log(`🎯 Space ID cible: ${spaceId}`);

    // Créer le Space sauf s'il existe déjà
    if (await spaceExists(spaceId)) {
      console.log(`⚠️  Le Space ${spaceId} existe déjà — mise à jour des fichiers.`);
    } else {
      await createSpace(spaceId, username, options.private);
    }

    // Upload des fichiers
    if (!options.skipFiles) {
      await uploadSpaceFiles(spaceId, sourceDir);
    }

    // Attendre le déploiement
    console.log('\n🔄 Le déploiement va prendre quelques minutes...');
    console.log(`   Vous pouvez suivre la progression sur:`);
    console.log(`   https://huggingface.co/spaces/${spaceId}`);

    const deployed = await waitForDeployment(spaceId);

    if (deployed) {
      console.log('\n✅ DÉPLOIEMENT TERMINÉ AVEC SUCCÈS!');
      console.log(`\n📍 URL du Space: https://huggingface.co/spaces/${spaceId}`);
      console.log(`\n💡 Endpoints utilisés par le Worker Cloudflare:`);
      console.log(`   - POST https://${spaceId.replace('/', '-')}.hf.space/api/convert-3d`);
      console.log(`   - POST https://${spaceId.replace('/', '-')}.hf.space/api/convert-office`);
      console.log(`\n⚠️  Note: Le Space se met en veille après inactivité.`);
      console.log(`   Premier appel = cold start (30-60 secondes)`);
    } else {
      console.log('\n⚠️  Déploiement en cours ou échoué - vérifiez les logs manuellement');
      console.log(`   https://huggingface.co/spaces/${spaceId}/tree/main`);
    }
  } catch (error) {
    console.error('\n❌ ERREUR CRITIQUE:', error.message);
    process.exit(1);
  }
}

// Exécution uniquement en appel direct (pas à l'import pour les tests)
const invokedAsScript = process.argv[1] && resolve(process.argv[1]) === fileURLToPath(import.meta.url);
if (invokedAsScript) {
  main().catch(console.error);
}

export { HF_API, SPACE_FILES, createSpace, parseArgs, spaceExists, uploadSpaceFiles, waitForDeployment };

File: scripts/deploy-space.py
#!/usr/bin/env python3
"""
Script de déploiement automatique du Space Hugging Face (version Python).

Utilisation:
    HF_TOKEN=your_token python scripts/deploy-space.py
    HF_TOKEN=your_token python scripts/deploy-space.py --private --skip-files

Variables d'environnement requises:
    - HF_TOKEN: Token Hugging Face avec permissions "write" et "repo.create"

Dépendance:
    pip install huggingface_hub

Note importante sur le plan Hugging Face:
    - Static Spaces : gratuits pour tout le monde (mais pas de backend Python).
    - Gradio/Docker Spaces (cpu-basic) : nécessitent un abonnement PRO.
      https://huggingface.co/pro
    - Comptes personnels gratuits : jusqu'à 2 Spaces Gradio sur ZeroGPU.
"""

from __future__ import annotations

import argparse
import os
import sys
import time
from pathlib import Path

try:
    from huggingface_hub import HfApi, SpaceStage
    from huggingface_hub.errors import HfHubHTTPError
except ImportError:
    print(
        "❌ La dépendance 'huggingface_hub' est manquante.\n"
        "   Installez-la avec:  pip install huggingface_hub\n"
    )
    sys.exit(1)

# ── Configuration ────────────────────────────────────────────────────────────
SPACE_NAME = "solidworks-viewer"
SPACE_SDK = "docker"
SPACE_HARDWARE = "cpu-basic"  # cpu-basic, cpu-upgrade, t4-medium, a10g-small, ...


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Déploiement automatique du Space SolidWorks Viewer (Hugging Face).",
    )
    parser.add_argument(
        "--space-id",
        default=None,
        help="ID du Space (format: username/space-name). Défaut: <username>/solidworks-viewer",
    )
    parser.add_argument(
        "--private",
        action="store_true",
        help="Créer un Space privé (défaut: public)",
    )
    parser.add_argument(
        "--skip-files",
        action="store_true",
        help="Ne pas uploader les fichiers (seulement la création du repo)",
    )
    parser.add_argument(
        "--skip-wait",
        action="store_true",
        help="Ne pas attendre la fin du déploiement",
    )
    parser.add_argument(
        "--sdk",
        default=SPACE_SDK,
        help=f"SDK du Space (défaut: {SPACE_SDK})",
    )
    parser.add_argument(
        "--hardware",
        default=SPACE_HARDWARE,
        help=f"Hardware du Space (défaut: {SPACE_HARDWARE})",
    )
    return parser.parse_args()


def get_username(api: HfApi) -> str:
    info = api.whoami()
    return info["name"]


def create_space(
    api: HfApi,
    space_id: str,
    is_private: bool,
    sdk: str,
    hardware: str,
) -> bool:
    """Crée le Space via huggingface_hub. Retourne True si créé, False s'il existe déjà."""
    print(f"\n🚀 Création du Space: {space_id}")
    print(f"   SDK: {sdk}")
    print(f"   Hardware: {hardware}")
    print(f"   Visibilité: {'privé' if is_private else 'public'}")

    if api.repo_exists(repo_id=space_id, repo_type="space"):
        print(f"⚠️  Le Space {space_id} existe déjà")
        return False

    api.create_repo(
        repo_id=space_id,
        repo_type="space",
        space_sdk=sdk,
        space_hardware=hardware,
        private=is_private,
        exist_ok=True,
    )

    print("✅ Space créé avec succès")
    return True


def upload_files(api: HfApi, space_id: str, source_dir: Path) -> None:
    """Upload tous les fichiers du dossier source vers la racine du Space."""
    print(f"\n📁 Upload des fichiers depuis {source_dir}")

    api.upload_folder(
        repo_id=space_id,
        repo_type="space",
        folder_path=str(source_dir),
        commit_message="Upload des fichiers du Space",
    )

    print("✅ Tous les fichiers ont été uploadés")


def wait_for_deployment(api: HfApi, space_id: str, timeout: int = 600) -> bool:
    """Attend que le Space passe en RUNNING. Timeout en secondes."""
    print(f"\n⏳ Attente du déploiement (timeout: {timeout // 60}min)...")

    start_time = time.time()
    check_interval = 10  # secondes

    while time.time() - start_time < timeout:
        try:
            runtime = api.get_space_runtime(space_id)
            stage = runtime.stage
            print(f"   Status: {stage}")

            if stage == SpaceStage.RUNNING:
                print("✅ Space déployé et opérationnel!")
                return True

            if stage == SpaceStage.RUNTIME_ERROR:
                print(f"❌ Erreur de déploiement: {runtime}")
                return False
        except HfHubHTTPError as exc:
            print(f"   Erreur vérification status: {exc}")

        time.sleep(check_interval)

    print("⏰ Timeout atteint - le déploiement est toujours en cours")
    return False


def main() -> None:
    args = parse_args()

    print("🔧 Déploiement automatique du Space SolidWorks Viewer\n")

    token = os.environ.get("HF_TOKEN")
    if not token:
        print("❌ Variable HF_TOKEN manquante")
        print("   Usage: HF_TOKEN=your_token python scripts/deploy-space.py")
        sys.exit(1)

    source_dir = Path(__file__).resolve().parent.parent / "space-huggingface"
    if not source_dir.is_dir():
        print(f"❌ Dossier source introuvable: {source_dir}")
        sys.exit(1)

    api = HfApi(token=token)

    try:
        # Récupérer le username
        print("📋 Récupération des informations utilisateur...")
        username = get_username(api)

        # Déterminer le spaceId
        space_id = args.space_id or f"{username}/{SPACE_NAME}"
        print(f"🎯 Space ID cible: {space_id}")

        # Créer le Space
        create_space(api, space_id, args.private, args.sdk, args.hardware)

        # Upload des fichiers
        if not args.skip_files:
            upload_files(api, space_id, source_dir)

        if args.skip_wait:
            print(f"\n📍 URL du Space: https://huggingface.co/spaces/{space_id}")
            return

        # Attendre le déploiement
        print("\n🔄 Le déploiement va prendre quelques minutes...")
        print(f"   Vous pouvez suivre la progression sur:")
        print(f"   https://huggingface.co/spaces/{space_id}")

        deployed = wait_for_deployment(api, space_id)

        if deployed:
            print("\n✅ DÉPLOIEMENT TERMINÉ AVEC SUCCÈS!")
            print(f"\n📍 URL du Space: https://huggingface.co/spaces/{space_id}")
            print("\n💡 Pour appeler ce Space depuis votre application:")
            print("   - Utilisez gradio_client (Python) ou @gradio/client (JS)")
            print(f"   - Endpoint: https://{space_id.replace('/', '-')}.hf.space")
            print("\n⚠️  Note: Le Space se met en veille après inactivité.")
            print("   Premier appel = cold start (30-60 secondes)")
        else:
            print("\n⚠️  Déploiement en cours ou échoué - vérifiez les logs manuellement")
            print(f"   https://huggingface.co/spaces/{space_id}/tree/main")

    except HfHubHTTPError as exc:
        status = getattr(getattr(exc, "response", None), "status_code", None)
        if status == 402:
            print("\n❌ ERREUR 402: votre compte nécessite un plan payant pour héberger ce Space.")
            print("   Le SDK choisi (Docker/Gradio) tourne sur du compute et requiert:")
            print("   - Un abonnement PRO (compte personnel): https://huggingface.co/pro")
            print("   - Ou Team/Enterprise (organisation).")
            print("\n   Alternatives gratuites:")
            print("   - Static Space (gratuit, mais pas de backend Python possible).")
            print("   - Jusqu'à 2 Spaces Gradio sur ZeroGPU pour un compte personnel gratuit.")
            print("     (nécessite de convertir le Space Docker → Gradio avec `packages.txt` pour FreeCAD)")
        else:
            print(f"\n❌ ERREUR API: {exc}")
        sys.exit(1)
    except Exception as exc:  # noqa: BLE001
        print(f"\n❌ ERREUR CRITIQUE: {exc}")
        sys.exit(1)


if __name__ == "__main__":
    main()
